prover/src/constraints/evaluator.rs

```diff
@@ -23,18 +26,117 @@ use utils::{iterators::*, rayon};
 #[cfg(feature = "concurrent")]
 const MIN_CONCURRENT_DOMAIN_SIZE: usize = 8192;
 
-// CONSTRAINT EVALUATOR
+// CONSTRAINT EVALUATOR TRAIT
 // ================================================================================================
 
-pub struct ConstraintEvaluator<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> {
+pub trait ConstraintEvaluator<'a, E: FieldElement> {
+    type Air: Air<BaseField = E::BaseField>;
+    type TraceLde: TraceLde<BaseField = E::BaseField>;
+    type HashFn: ElementHasher<BaseField = E::BaseField>;
+
+    fn evaluate(
+        self,
+        trace: &Self::TraceLde,
+        domain: &'a StarkDomain<E::BaseField>,
+    ) -> ConstraintEvaluationTable<'a, E>;
+}
+
+// CONSTRAINT EVALUATOR DEFAULT IMPLEMENTATION
+// ================================================================================================
+
+pub struct DefaultConstraintEvaluator<
+    'a,
+    A: Air,
+    E: FieldElement<BaseField = A::BaseField>,
+    H: ElementHasher<BaseField = E::BaseField>,
+> {
     air: &'a A,
     boundary_constraints: BoundaryConstraints<E>,
     transition_constraints: TransitionConstraints<E>,
     aux_rand_elements: AuxTraceRandElements<E>,
     periodic_values: PeriodicValueTable<E::BaseField>,
+    hasher: PhantomData<H>,
 }
 
-impl<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> ConstraintEvaluator<'a, A, E> {
+impl<
+        'a,
+        A: Air,
+        E: FieldElement<BaseField = A::BaseField>,
+        H: ElementHasher<BaseField = E::BaseField> + std::marker::Sync,
+    > ConstraintEvaluator<'a, E> for DefaultConstraintEvaluator<'a, A, E, H>
+{
```

Comment:
Assuming we change the definition of the trait as described above, we can change this to:
```Rust
impl<'a, A, E> ConstraintEvaluator<'a, E> for DefaultConstraintEvaluator<'a, A, E>
where
    A: Air,
    E: FieldElement<BaseField = A::BaseField>,
{
```